tools/publicapigen/main.go

```diff
@@ -0,0 +1,463 @@
+					}
+				}
+
+			case *ast.ValueSpec:
+				// Remove unexported variables and constants
+				keep := false
+				for i, name := range node.Names {
+					if name.IsExported() {
+						keep = true
+					} else {
+						node.Names[i] = ast.NewIdent("_")
+					}
+				}
+
+				if mustKeep(node.Doc, node.Comment) {
+					keep = true
+				}
+
+				if !keep {
+					c.Delete()
+					return false
+				}
+
+			case *ast.SelectorExpr:
+				// else if we have a selector, rewrite the package name to see if this is an export from
+				// one of our private packages, in which case we want to copy it into the public API
+				if pkg, ok := node.X.(*ast.Ident); ok && constants[pkg.Name] != nil {
+					typ, found := constants[pkg.Name][node.Sel.Name]
+					if found {
+						typ.ValuePos = token.NoPos
+						c.Replace(&typ)
+					}
+				}
+			case *ast.GenDecl:
+				if mustKeep(node.Doc) {
+					return false
+				}
+
+			case *ast.Field:
+				// Keep exported fields if there's a doc saying we should
+				keep := mustKeep(node.Doc, node.Comment)
+
+				if !keep {
+					// Remove unexported fields from structs
+					for i, name := range node.Names {
+						if name.IsExported() {
+							keep = true
+						} else {
+							node.Names[i] = ast.NewIdent("_")
+						}
+					}
+				}
+
+				if !keep {
+					c.Delete()
+				}
+				return false
+			}
+
+			return true
+		},
+		func(c *astutil.Cursor) bool {
+			switch node := c.Node().(type) {
+			case *ast.GenDecl:
+				if len(node.Specs) == 0 {
+					clearCommentGroup(node.Doc)
+					c.Delete()
+				}
+
+			case *ast.StructType:
+				if len(node.Fields.List) == 0 {
+					node.Fields.List = append(node.Fields.List, &ast.Field{
+						Doc:   nil,
+						Names: []*ast.Ident{ast.NewIdent("_")},
+						Type:  ast.NewIdent("int"),
+						Tag:   nil,
+						Comment: &ast.CommentGroup{List: []*ast.Comment{
+							{
+								Slash: token.NoPos,
+								Text:  "// for godoc to show unexported fields",
+							},
+						}},
+					})
+				}
+			}
+
+			return true
+		},
+	)
+	return nil
+}
+
+// walkDir recursively descends path, calling walkFn for directory
+func walkDir(dir, rel string, f func(path, rel string, files []os.FileInfo) error) error {
+	if rel == "types/uuid" {
+		// we don't want to rewrite this package
+		return nil
+	}
+
+	log.Debug().Str("rel", rel).Msg("Walking directory")
+	entries, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return err
+	}
+
+	// Split the files and dirs
+	var dirs, files []os.FileInfo
+	for _, entry := range entries {
+		if entry.IsDir() {
+			dirs = append(dirs, entry)
+		} else {
+			files = append(files, entry)
+		}
+	}
+
+	if err := f(dir, rel, files); err != nil {
+		return fmt.Errorf("error walking %s: %w", rel, err)
+	}
+
+	for _, d := range dirs {
+		dir2 := filepath.Join(dir, d.Name())
+		rel2 := path.Join(rel, d.Name())
+
+		if err := walkDir(dir2, rel2, f); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func repoDir() string {
+	_, file, _, ok := runtime.Caller(0)
```

Comment:
The output of `runtime.Caller(0)` will involve file paths based on where the binary was built, which means this could cause surprising results if run on someone else's machine. I guess that's quite unlikely and it doesn't look like we're doing anything bad with the return value (if we called `os.RemoveAll` or something that could be bad)